pkg/promtail/targets/lokipush/pushtarget.go

```diff
@@ -0,0 +1,203 @@
+package lokipush
+
+import (
+	"flag"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/imdario/mergo"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/relabel"
+	"github.com/weaveworks/common/server"
+
+	"github.com/grafana/loki/pkg/distributor"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/pkg/promtail/targets/target"
+)
+
+type PushTarget struct {
+	logger        log.Logger
+	handler       api.EntryHandler
+	config        *scrapeconfig.PushTargetConfig
+	relabelConfig []*relabel.Config
+	jobName       string
+	server        *server.Server
+}
+
+func NewPushTarget(logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.PushTargetConfig) (*PushTarget, error) {
+
+	pt := &PushTarget{
+		logger:        logger,
+		handler:       handler,
+		relabelConfig: relabel,
+		jobName:       jobName,
+		config:        config,
+	}
+
+	// Bit of a chicken and egg problem trying to register the defaults and apply overrides from the loaded config.
+	// First create an empty config and set defaults.
+	defaults := server.Config{}
+	defaults.RegisterFlags(flag.NewFlagSet("empty", flag.ContinueOnError))
+	// Then apply any config values loaded as overrides to the defaults.
+	if err := mergo.Merge(&defaults, config.Server, mergo.WithOverride); err != nil {
+		level.Error(logger).Log("msg", "failed to parse configs and override defaults when configuring push server", "err", err)
+	}
+	// The merge won't overwrite with a zero value but in the case of ports 0 value
+	// indicates the desire for a random port so reset these to zero if the incoming config val is 0
+	if config.Server.HTTPListenPort == 0 {
+		defaults.HTTPListenPort = 0
+	}
+	if config.Server.GRPCListenPort == 0 {
+		defaults.GRPCListenPort = 0
+	}
+	// Set the config to the new combined config.
+	config.Server = defaults
+
+	err := pt.run()
+	if err != nil {
+		return nil, err
+	}
+
+	return pt, nil
+}
+
+func (t *PushTarget) run() error {
+	level.Info(t.logger).Log("msg", "starting push server", "job", t.jobName)
+	// To prevent metric collisions because all metrics are going to be registered in the global Prometheus registry.
+	t.config.Server.MetricsNamespace = "promtail_" + t.jobName
+
+	// We don't want the /debug and /metrics endpoints running
+	t.config.Server.RegisterInstrumentation = false
+
+	util.InitLogger(&t.config.Server)
+
+	srv, err := server.New(t.config.Server)
+	if err != nil {
+		return err
+	}
+
+	t.server = srv
+	t.server.HTTP.Handle("/loki/api/v1/push", http.HandlerFunc(t.handle))
+
+	go func() {
+		err := srv.Run()
```

Comment:
We wouldn't be able to block and wait for the error because the `Run` method blocks forever when the server runs normally.  

The creation of the server object itself actually opens the port and starts the server so any errors there would def get caught and propagated up and would prevent promtail from starting.

I think logging here is probably the best we can do if for some reason the server stops unexpectedly after it was started.